[SPARK-44059] Add analyzer support of named arguments for built-in functions

learningchess2003 opened the pull request:
### What changes were proposed in this pull request?
Add analyzer support for named function arguments. 

### Why are the changes needed?
Part of the project needed for general named function argument support.

### Does this PR introduce _any_ user-facing change?
We added support for named arguments for the ```CountMinSketchAgg``` and ```Mask``` SQL functions.

### How was this patch tested?
A new suite was added for this test called NamedArgumentFunctionSuite.

learningchess2003 committed:
SPARK-44059

learningchess2003 committed:
Some weird files shouldn't be in here

learningchess2003 commented:
@dtenedor @cloud-fan @anchovYu @MaxGekk  @allisonwang-db Please take a look when you have time!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
+  usage = "_FUNC_(expr) - Separates the elements of array `expr` into multiple rows, or the elements of map `expr` into multiple rows and columns. Unless specified otherwise, uses the default column name `col` for elements of the array or `key` and `value` for the elements of the map.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(10, 20));
```
This is the function doc of the TVF version of `explode`. Shall we show the example with the TVF syntax? `SELECT * FROM _FUNC_(array(10, 20))`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
+  usage = "_FUNC_(expr) - Separates the elements of array `expr` into multiple rows, or the elements of map `expr` into multiple rows and columns. Unless specified otherwise, uses the default column name `col` for elements of the array or `key` and `value` for the elements of the map.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(10, 20));
```
ditto

dtenedor reviewed (approved):
LGTM 👍

learningchess2003 committed:
Addressing nits

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
+      > SELECT _FUNC_(array(10, 20));
+       10
+       20
+      > SELECT _FUNC_(collection => array(10, 20));
```
Can we also add one example for `pos_explode`?

allisonwang-db reviewed (commented):
Looks good!

anchovYu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/FunctionBuilderBase.scala:
```diff
@@ -0,0 +1,179 @@
+ * @param default  The default value of the argument. If the default is none, then that means the
+ *                 argument is required. If no argument is provided, an exception is thrown.
+ */
+case class NamedArgument(
```
nit: Can this case class fit on the same line?

anchovYu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:
```diff
@@ -75,6 +76,25 @@ import org.apache.spark.unsafe.types.UTF8String
+  }
+
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    new Mask(expressions(0), expressions(1), expressions(2), expressions(3), expressions(4))
```
nit: Could you still add an assertion on the size of the expressions? This way `build` here as an independent function is more complete itself.

anchovYu commented on common/utils/src/main/resources/error/error-classes.json:
```diff
@@ -1956,7 +1974,13 @@
     ]
   },
-  "NAMED_ARGUMENTS_SUPPORT_DISABLED" : {
+  "NAMED_PARAMETERS_NOT_SUPPORTED" : {
```
nit: The object is named as `object NamedArgumentsSupport`. Could you unify the naming on the 'parameters' and 'arguments'?

anchovYu replied:
nvm, it seems reasonable for the naming of `NamedArgumentsSupport ` but `NAMED_PARAMETERS_NOT_SUPPORTED`.

learningchess2003 replied:
Serge was actually the one who recommended the name change. It should be fine.

anchovYu replied:
Right, I agree with the change in the error class to parameter, but I was thinking if similar changes should be applied on `NamedArgumentsSupport`, e.g. `NamedParametersSupport`.

learningchess2003 replied:
Okay, that makes sense.

anchovYu reviewed (approved):
lgtm!

learningchess2003 committed:
Adding some documentation

learningchess2003 committed:
Making consistent names

learningchess2003 commented:
@cloud-fan Thanks for all the comments! I've addressed comments from the other reviewers. All of the tests passed except a flaky which I might rerun to see if we can pass it. Otherwise, I've already seen an instance where we had an all green.  If you think this PR is ready, you can merge it.

cloud-fan commented:
thanks, merging to master/3.5! (half of this feature is already in 3.5)
